### PR TITLE
Added support for python scripts. #1415

### PR DESCRIPTION
**Pull request by trevor-james-nangosha**
Regarding issue #1415, I have added support for Python scripts by checking for a shebang at the start of the file with the path to the Python executable.

One thing to note with this approach is that it will check all files in the directory/those that you provide to see if they are python executable scripts. That is, even if you have a file called `main.html`, but it has a python shebang line at the top and has python code, then it will be type checked.

This works for me, but I am also looking forward to any opinions that could provide more robust solutions.

### CODE REVIEW

**Comment by rchen152** [comment]
Sorry for the slow response! This looks like it would open and read every file in a directory and subdirectories to find Python scripts? That seems like it would be too slow to be practical, unfortunately.

What do you think of only allowing directly specified files to have non-.py extensions? That is, if you run `pytype my_script`, then `my_script` should be analyzed, but pytype will still only search for `.py` files when looking for dependencies of `my_script`.

**Comment by trevor-james-nangosha** [comment]
@rchen152. That makes sense. I'll get on it right away.

**Comment by trevor-james-nangosha** [comment]
@rchen152, have added another commit to only check for python scripts when a non-`.py` file is specified.

File: pytype/file_utils.py [hunk 1]
```diff
@@ -136,3 +140,12 @@ def replace_separator(path: str):
...
     return path
+  
+def is_file_script(filename, dir=None):
+  # This is for python files that do not have the .py extension
+  # of course we assume that they start with a shebang
+  file_path = expand_path(filename, dir)
+  if path_utils.isfile(file_path):
+    with open(file_path, 'r') as file:
+      line = file.readline().rstrip().lower()
+      return line.startswith('#!') and (line.endswith('python') or line.endswith('python3'))
```
Comment (rchen152) [inline pytype/file_utils.py:151]: I'd recommend using a regex here. Something like `return re.fullmatch(r'#!.+python3?', line) is not None` should work.

File: pytype/file_utils.py [hunk 2]
```diff
@@ -114,10 +114,14 @@ def expand_source_files(filenames, cwd=None):
   out = []
   for f in expand_globpaths(filenames.split(), cwd):
     if path_utils.isdir(f):
-      # If we have a directory, collect all the .py files within it.
-      out += recursive_glob(path_utils.join(f, "**", "*.py"))
+      # If we have a directory, collect all the .py files within it....
+      out  += recursive_glob(path_utils.join(f, "**", "*.py"))
```
Comment (rchen152) [inline pytype/file_utils.py:118]:
```suggestion
      out += recursive_glob(path_utils.join(f, "**", "*.py"))
```

File: pytype/file_utils_test.py [hunk 3]
```diff
@@ -107,9 +107,15 @@ class TestExpandSourceFiles(unittest.TestCase):
...
 
+  SCRIPT_CODE = """
+  #!/usr/bin/python3
+  print("hello, world")
+  """
+
   def _test_expand(self, string):
     with test_utils.Tempdir() as d:
       fs = [d.create_file(f) for f in self.FILES]
+      fs += [d.create_file("my_script", self.SCRIPT_CODE)]
```
Comment (rchen152) [inline pytype/file_utils_test.py:118]: What is this line accomplishing? The test still asserts that expand_source_files only finds the .py files.

**Review by rchen152 — CHANGES_REQUESTED** [review]
Thanks! A few comments.

File: pytype/file_utils_test.py (hunk 3 above)
Comment (trevor-james-nangosha) [reply pytype/file_utils_test.py:118]: Previously when I was checking every file in the directory to check if it was a script, this line would have made sense to make sure we don't miss a script file. Now that we only check for a file that is passed to `pytype`, this line  looks redundant.

I will remove it. Thanks for pointing that out.

**Review by rchen152 — CHANGES_REQUESTED** [review]
Looking good! There are a couple of lint errors: https://github.com/google/pytype/actions/runs/4945740447/jobs/8861864375

- There's extra whitespace on line 144.
- The parameter name `dir` on line 145 conflicts with the `dir` builtin.

Once those are fixed, I think this will be ready to merge.

**Comment by trevor-james-nangosha** [comment]
@rchen152, Done!!!!!!!!

**Pull request merged**